circuit/trade/trade.func.circom

```diff
@@ -0,0 +1,82 @@
+pragma circom 2.1.5;
+
+include "../node_modules/circomlib/circuits/bitify.circom";
+include "../node_modules/circomlib/circuits/comparators.circom";
+include "../node_modules/circomlib/circuits/poseidon.circom";
+
+function u128_ceiling() {
+    return 340282366920938463463374607431768211455;
+}
+
+template Trade(n) {
+    // =========================================
+    // Private Inputs
+    // =========================================
+    signal input amount_in;
+    signal input reserve_in;
+    signal input reserve_out;
+    signal input amount_out;
+    signal input salt;
+
+    // =========================================
+    // Public Inputs
+    // =========================================
+    signal input reserve_in_image;
+    signal input reserve_out_image;
+
+    // =========================================
+    // Verify preimage knowledge
+    // =========================================
+    component hasher_reserve_in = Poseidon(2);
```

Comment:
preference for anonymous components for readability

https://docs.circom.io/circom-language/anonymous-components-and-tuples/#:~:text=circom%202.1.,assignment%20of%20every%20output%20signal.